### PR TITLE
Move more pass from Flow stage to GlobalOptimization stage.

### PR DESCRIPTION
**Pull request by hanhanW**
- Move four more passes to GlobalOptimization stage.
  - ConvertElementwiseToLinalgPass
  - GeneralizeLinalgNamedOpsPass
  - FuseDequantizationMatmulPass
  - FoldUnitExtentDimsPass
 - Move Flow transformation_pipeline.mlir test to GlobalOptimization/test. It is mainly for testing ConvertElementwiseToLinalg pass which is also tested upstream. We probably can remove it as a follow-up.

### CODE REVIEW

**Review by benvanik — APPROVED** [review]

**Comment by hanhanW** [comment]
The result is interesting... overall is positive to me.

**Comment by MaheshRavishankar** [comment]
I'd try to triage the difference in the number of dispatches created.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
I think needs a bit of triage on the number of dispatches created.

**Comment by hanhanW** [comment]
Agree that we need more investigation. I was just trying to see what's happening in this case. I'd like to scope it moving SetEncoding to FlowPreprocessing stage. (The invesgitation can happen when we work on improving const-eval heuristic. At that point we need to do some basic fusion and move some passes to preprocessing stage.)

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp [hunk 1]
```diff
@@ -214,12 +214,13 @@ void buildFlowTransformPassPipeline(OpPassManager &passManager,
...
       .addPass(IREE::Flow::createRemoveZeroExtentTensorsPass)
       .addPass(IREE::Flow::createDetachElementwiseFromNamedOpsPass)
       .addPass(mlir::createLinalgNamedOpConversionPass)
-      .addPass(IREE::Flow::createConvert1X1FilterConv2DToMatmulPass);
-  passManager.addPass(IREE::Flow::createEraseUnusedLinalgOperands());
+      .addPass(IREE::Flow::createConvert1X1FilterConv2DToMatmulPass)
+      .addPredicatedPass(clEnableDataTiling, createSetEncodingPass);
 
   // Start of Flow pipeline, verify input legality.
   passManager.addPass(IREE::Flow::createVerifyInputLegalityPass());
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp]: Do you want to move this also before the preprocessing passes?

**Comment by hanhanW** [comment]
@MaheshRavishankar We can't move raising special op pass before global optimization now. Because it introduces huge regression to CUDA. Something falls into sequential computation. So far we can move few passes to global optimization phase. Please take a look if it is okay to move them to global optimization phase. It helps us enable const-eval for data tiling (i.e., https://github.com/openxla/iree/pull/14792) because we wont need to handle rank-reduced cases. We can run FoldUnitExtentDimsPass pass before SetEncoding pass.

**Comment by benvanik** [comment]
(wonder if this is CUDA transform dialect matchers or some other CUDA-specific patterns special cased on rank or something?)

**Comment by hanhanW** [comment]
> (wonder if this is CUDA transform dialect matchers or some other CUDA-specific patterns special cased on rank or something?)

I don't know at this moment. :) My take is that RaiseSpecialOps pass is a preprocessing for fusion. If we move it before const-eval, const-eval could break the behavior. It hoists some ops to globals, which breaks the graph. So the matcher no longer work. It is one of passes that should be run right before fusion.

**Comment by qedawkins** [comment]
Can we run RaiseSpecialOps in multiple places? I could see it being worth running both before and after FoldUnitExtentDims/other passes.

**Comment by benvanik** [comment]
RaiseSpecialOps indeed would be useful to run at various stages, including possibly as part of a fixed point iteration around many such passes.

**Comment by qedawkins** [comment]
Cool, that matches my thinking about the pass as well. It should basically contain patterns that are always worth applying, but we don't know exactly when they're going to apply.

**Comment by MaheshRavishankar** [comment]
I'd rather wait for thigns to stabilize before we run them multiple times.....

**Review by MaheshRavishankar — APPROVED** [review]
Ok, this is fine for now.

**Comment by hanhanW** [comment]
I think we can run it multiple times, if we do know where we want to apply them. I'm not going to study if we can put it to multiple places or not in this PR. The intention is moving more passes to global optimization phase if it makes sense.

**Comment by stellaraccident** [comment]
Note that this increased the amount of constant folding by 20-30x on llama2: it is memorizing a bunch of collapse_shapes mostly, it seems. We may want to deny-list collaspe_shape as constexpr leaf nodes because there is seldom any real value in constevaling a metadata change. 

Does not seem to have made any significant change to latency, either of runtime or compile time (except for one compile time outlier that I assume was a fluke).

**Pull request merged**